Catalogue : Add UI to manage columns

themissingcow opened the pull request:
![image](https://user-images.githubusercontent.com/896779/77537790-95776100-6e96-11ea-9dfd-58e1e1fca13d.png)

- Adds UI to control column visibility via right-click on the header.
- Adds context variables to image header under the `gaffer:context:` prefix.
- Adds additional column presets.

NB: This is a 'minimal' UI, that allows order customisation via the order you add columns. Full drag-drop etc. is a bit of a faf right now with `PathListingWidget` so it'd be good to save this until it becomes completely essential...

andrewkaufman commented on src/GafferScene/RendererAlgo.cpp:
```diff
@@ -1274,6 +1275,17 @@ ConstOutputPtr addGafferOutputHeaders( const Output *output, const ScenePlug *sc
+	{
+		// Output params are never mutated, so this is 'safe'...
+		DataPtr data = const_cast<Data*>( context->get<Data>( name ) );
+		param->writable()["header:gaffer:context:" + name.string()] = data;
```
I'm getting a handful of warnings related to this

```
WARNING : IECoreArnold::Renderer : Cannot convert data "gaffer:context:image:defaultFormat" of type "GafferImage::FormatData".
WARNING : setParameter : Unsupported data type "GafferImage::FormatData" for name "header:gaffer:context:image:defaultFormat"
WARNING : getParameters : Unable to convert user parameter "header:gaffer:context:defaultResolution"
WARNING : getParameters : Unable to convert user parameter "header:gaffer:context:defaultOverscan"
```

The defaultResolution & defaultOverscan are both V2i context entries.

johnhaddon replied:
We'll deal with this for now by using a whitelist of types we think a renderer can reasonably be expected to support. Tom, please also check that we don't get warnings when rendering to EXR via Arnold, and also not when rendering to Appleseed (both IPR and batch).

andrewkaufman replied:
To clarify, are V2i on the whitelist or the blacklist? It feels like a pretty basic type to me.

themissingcow replied:
For now, it'll be whatever is supported by the various render backends. For this to work, they have to be convertible both to and then from the renderers native types. I think it's out of scope for this PR to get into extending what we already have.

themissingcow replied:
https://github.com/ImageEngine/cortex/blob/master/contrib/IECoreArnold/src/IECoreArnold/ParameterAlgo.cpp#L270 is quite limited at present

themissingcow replied:
The warnings should be resolved in https://github.com/GafferHQ/gaffer/pull/3688/commits/b8b7d43a558b865dd10532c082837b22bf6e37cb via whitelisting of types. If we need to support more, let us know and we can address in a subsequent PR.

andrewkaufman commented:
Also, I would expect my column layout to serialize with the node, currently a copy/paste of my IPR Box loses those columns, as does a scene save / reopen.

johnhaddon commented on startup/GafferSceneUI/catalogue.py:
```diff
@@ -67,8 +67,16 @@ def typeIconColumnValueProvider( image, catalogue ):
+
+def renderNodeColumnValueProvider( image, catalogue ):
+	scenePlug = GafferScene.SceneAlgo.sourceScene( catalogue["out"] )
+	return scenePlug.node().getName() if scenePlug else ""
```
I think the right approach here is to traverse up the hierarchy until there is no ancestor for which the "graphEditor:childrenViewable" metadata is False.

Alex brought up a related problem very recently too - SourceSet and SceneHistoryUI use `ancestorWithReadOnlyChildNodes()` to determine the first publicly viewable ancestor, but they overdo it by traversing out of References when it would be more useful if they showed the node within. The same "graphEditor:childrenViewable" approach would fix that.

Let's use this opportunity to kill two birds with one stone. Should we perhaps consider a more generic name than "graphEditor:childrenViewable", and add a method in MetadataAlgo to find the appropriate ancestor?

themissingcow replied:
I've dropped this column for now. We can pick that back up again in the near future if people determine `Render Node` is a useful thing. It wasn't on the original list, I just added it as we had the header around.

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -150,6 +150,7 @@ def __statusIconColumnValueProvider( image, catalogue ) :
 registerColumn( "status", "", __statusIconColumnValueProvider, ColumnType.Icon )
 registerColumn( "name", "Name", lambda image, _ : image.getName() )
 registerColumn( "frame", "Frame", contextValueProvider( "frame" ) )
+registerColumn( "description", "Description", lambda image, _ : image["description"].getValue() )
```
`image["description"]` feeds into `catalogue["out"].metadata()["ImageDescription"]`. We would be better off reading from the latter, because that works for images that are loaded from disk too.

themissingcow replied:
Updated in https://github.com/GafferHQ/gaffer/pull/3688/commits/23cef312a7dba8351bf00bbdf68450d3e2ad12d6

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -880,4 +910,58 @@ def __keyPress( self, imageListing, keyEvent ) :
+		for column in allColumnsSorted :
+
+			definition = columnDefinition( column )
+			path = "/".join( [ IECore.CamelCase.toSpaced(p) for p in column.split( "/") ])
```
Let's use the name verbatim, and register with the exact user-visible names we want in the first place. That's what we do for presets and outputs and so on...

themissingcow replied:
Fixed https://github.com/GafferHQ/gaffer/pull/3688/commits/78ba11e3bde4ea782b13c5a708537b676d05cb8b#diff-f8b21d2701a672983fe238345811269bR979, and in all other column registrations.

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -103,19 +105,90 @@ def registeredColumns() :
+	value = "%d x %d" % ( format_.width(), format_.height() )
+	ratio = format_.getPixelAspect()
+	if ratio != 1.0 :
+		value += " [%d x %d]" % ( format_.width() * ratio, format_.height() )
```
Agreed. Perhaps pixel aspect should just be a separate column?

johnhaddon replied:
Oh, I see it is a separate column already. Let's just drop it from here then.

themissingcow replied:
It doesn't list pixel aspect here, it just shows the resulting resolution, which seemed like it might be useful. @andrewkaufman I tried generally speaking to match how we present it, but currently, we do things differently between 2D and 3D views anyway ( eg: `XxY` and `X x Y` ). We should unify those at some point.

themissingcow replied:
Removed in  https://github.com/GafferHQ/gaffer/pull/3688/commits/6c91dc3b6c9acfbb3dd02050157c9abb8f52992a#diff-f8b21d2701a672983fe238345811269bR195-R198

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -152,6 +152,40 @@ def __statusIconColumnValueProvider( image, catalogue ) :
+registerColumn( "image/resolution", "Resolution", __resolutionColumnValueProvider )
+registerColumn( "image/channels", "Channels", __channelsColumnValueProvider )
+
+registerColumn( "image/pixelAspectRatio", "Pixel AR",
```
Maybe "Pixel AR" -> "Aspect"? Or "P. Aspect"? Isn't AR some new fangled thing the kids are using these days?

themissingcow replied:
Renamed to `P. Aspect` in https://github.com/GafferHQ/gaffer/pull/3688/commits/6c91dc3b6c9acfbb3dd02050157c9abb8f52992a#diff-f8b21d2701a672983fe238345811269bR218 to avoid confusion with the image's aspect ratio.

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -103,6 +103,41 @@ def registeredColumns() :
+# Returns a column value provider that retrieves a render context value from
+# the catalogue items image.  If multiple names are provided, the first present
+# context entry will be used.
+def contextValueProvider( nameOrNames, defaultValue = None ) :
```
Alternatively, perhaps we should follow the example of PathListingWidget and have a proper Column base class with ContextColumn and MetadataColumn derived classes. This would also address the fact that you didn't want to pass columns directly to `registerColumn()` because named tuples weren't very nice to use.

themissingcow replied:
I guess I liked de-coupling the fact that the Catalogue uses a `PathLIstingWidget`  internally, given we're talking about a lot of changes there.

themissingcow replied:
https://github.com/GafferHQ/gaffer/pull/3688/commits/de91b30ea03e6e24f1b55a52ac5d4d6086dc6b1e, https://github.com/GafferHQ/gaffer/pull/3688/commits/3835bac98d0a9f594273a44f4d0b49ce79d04e15 et. al. refactor to the Class approach.

themissingcow commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -844,4 +945,58 @@ def __keyPress( self, imageListing, keyEvent ) :
+		# This signal is called anywhere in the listing, check we're over the header.
+		mousePosition = GafferUI.Widget.mousePosition( relativeTo = self.__pathListing )
+		headerRect = self.__pathListing._qtWidget().header().rect()
+		if not headerRect.contains( mousePosition[0], mousePosition[1] ) :
```
I'd rather we got async render launching etc. working first, unless its a real deal breaker for people.

johnhaddon replied:
Agreed. It'd be a useful feature, but I think this is good enough for a first version. We have lots of other fish to fry...

themissingcow commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -880,4 +910,58 @@ def __keyPress( self, imageListing, keyEvent ) :
 
+	def __columnContextMenuDefinition( self ) :
+
+		weakSelf = weakref.ref( self )
```
As it'd need two weak methods rather than one weak self?

themissingcow replied:
Moved to instance method in https://github.com/GafferHQ/gaffer/pull/3688/commits/78ba11e3bde4ea782b13c5a708537b676d05cb8b#diff-f8b21d2701a672983fe238345811269bR629-R638

murrays-ie commented:
Nice stuff Tom! This is going to be extremely useful once finished. One small thing I've noticed in a quick test of the current CI build is that the frame column doesn't update if you start an IPR and then change the frame, so it leaves you thinking the render was of frame 1001 when actually you were last on 1011. At IE not such a big deal as we end up generating a new image whenever the frame is changed, but for regular Gafferites this could be an issue...

themissingcow commented:
Hey @murrays-ie, thanks for checking it out, glad it's looking useful.

>  One small thing I've noticed in a quick test of the current CI build is that the frame column doesn't update if you start an IPR and then change the frame

Sadly, it's a limitation at the moment.  We can only show headers as they are at the start of the render. We don't have a way to update them as the render progresses. So all catalogue data (aside from name/description/filename) will be as it was at the beginning, rather than end.  It's something we can fix, its just a disproportionate amount of work at the mo. Let us know if it turns out to be something we might more generally need.

themissingcow commented:
> Also, I would expect my column layout to serialize with the node, currently a copy/paste of my IPR Box loses those columns, as does a scene save / reopen.

Thanks for catching that Andrew. The revised PR moves this to plug metadata, and fixes a bunch of bugs with promoted plugs. https://github.com/GafferHQ/gaffer/pull/3688/commits/78ba11e3bde4ea782b13c5a708537b676d05cb8b#diff-f8b21d2701a672983fe238345811269bR609-R627

themissingcow commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -103,19 +105,90 @@ def registeredColumns() :
+		metadata = catalogue["out"].metadata()
+		for name in nameOrNames :
+			if name in metadata :
+				return metadata[ name ]
```
https://github.com/GafferHQ/gaffer/pull/3688/commits/3835bac98d0a9f594273a44f4d0b49ce79d04e15#diff-f8b21d2701a672983fe238345811269bR154-R155

themissingcow commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -103,19 +105,90 @@ def registeredColumns() :
+# catalogue items image.  If multiple names are provided, the first present in
+# the image headers will be used, allowing a single column to support several
+# source headers depending on the image's origin.
+def imageHeaderValueProvider( nameOrNames, defaultValue = None ) :
```
https://github.com/GafferHQ/gaffer/pull/3688/commits/3835bac98d0a9f594273a44f4d0b49ce79d04e15 renames to metadata

themissingcow commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -150,8 +152,8 @@ def __typeIconColumnValueProvider( image, catalogue ) :
 Gaffer.Metadata.registerValue(
-	GafferImage.Catalogue, "catalogue:columns",
 	IECore.StringVectorData( [ "typeIcon", "name" ] )
+	GafferImage.Catalogue, _columnsMetadataKey,
```
Good spot, thanks. The intermediate brokenness should be fixed in the new rebase after the refactor.

themissingcow commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -461,7 +463,7 @@ def __init__( self, plug, **kw ) :
 		with self.__column :
 
-			columns = self.__listingColumns()
+			columns = self.__pathWidgetColumns()
```
Reverted in  https://github.com/GafferHQ/gaffer/pull/3688/commits/78ba11e3bde4ea782b13c5a708537b676d05cb8b

themissingcow commented on python/GafferSceneTest/InteractiveRenderTest.py:
```diff
@@ -98,6 +98,8 @@ def testOutputs( self ):
 	def testMetadata( self ) :
 
 		s = Gaffer.ScriptNode()
+		s["variables"].addChild( Gaffer.NameValuePlug( "a", IECore.StringData( "A" ), "member1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
```
Ahh cool, I'd borrowed from some other test somewhere, thanks. Updated in https://github.com/GafferHQ/gaffer/pull/3688/commits/b8b7d43a558b865dd10532c082837b22bf6e37cb#diff-5f4989453d4dbfddd684c9113f35453fR101

themissingcow commented:
@andrewkaufman @johnhaddon (as agreed) have addressed these comments as a re-base due to the refactor. Andrew, _hopefully_ this should fix all your promoted woes.

The remaining issue of a horizontal scroll bar will be addressed in a separate PR due to some underlying bugs in the stylesheet(?) that causes the `AsNeccesary` scroll bar mode not to work correctly.

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -190,6 +190,44 @@ def value( self, image, catalogue ) :
+)
+registerColumn(
+	"Image/Pixel Type",
+	SimpleColumn( "Pixel Type", lambda _, c : "Deep" if c["out"].deep() else "Flat" )
```
Pixel Type -> Image Type perhaps? When I saw "Pixel Type" I thought int/half/float, even though I know Gaffer only deals in floats.

johnhaddon commented on python/GafferImageUITest/CatalogueUITest.py:
```diff
@@ -67,9 +67,34 @@ def testBoxedCatalogue( self ) :
+		sw = GafferUI.ScriptWindow.acquire( s )
+
+		with IECore.CapturingMessageHandler() as mh :
+			ne = GafferUI.NodeEditor.acquire( s["c"] )
```
Do we need the ScriptNode, ScriptWindow and NodeEditor here, rather than just using `NodeUI.create( catalogue )` directly? If possible, let's go with the simpler option and keep the test scoped more tightly to the bits we care about.

johnhaddon replied:
Due to complications arising from lazy UI creation on 0.56_maintenance, we're going to keep this as it is for now...
